dftimewolf/lib/collectors/grr_hosts.py

```diff
@@ -980,64 +981,85 @@ def _DownloadResults(self,
+          query=query,
+          hostname=hostname,
+          data_frame=data_frame,
+          flow_identifier=flow_identifier,
+          client_identifier=client_identifier)
+
+      self.state.StoreContainer(dataframe_container)
+
   def Process(self, container: containers.Host) -> None:
     """Collect osquery results from a host with GRR.
 
     Raises:
       DFTimewolfError: if no artifacts specified nor resolved by platform.
     """
-    for client in self._FindClients([container.hostname]):
-      osquery_containers = self.state.GetContainers(containers.OsqueryQuery)
+    osquery_containers = self.state.GetContainers(containers.OsqueryQuery)
 
+    host_osquery_futures = []
+    with ThreadPoolExecutor() as executor:
       for osquery_container in osquery_containers:
-        hunt_args = osquery_flows.OsqueryFlowArgs(
-            query=osquery_container.query,
-            timeout_millis=self.timeout_millis,
-            ignore_stderr_errors=self.ignore_stderr_errors)
-
-        try:
-          flow_id = self._LaunchFlow(client, 'OsqueryFlow', hunt_args)
-          self._AwaitFlow(client, flow_id)
-        except DFTimewolfError as error:
-          self.ModuleError(
-              f'Error raised while launching/awaiting flow: {error.message}')
-          continue
-
-        name = osquery_container.name
-        description = osquery_container.description
-        query = osquery_container.query
-        hostname = container.hostname
-        flow_identifier = flow_id
-        client_identifier = client.client_id
-
-        results = self._DownloadResults(client, flow_id)
-        if not results:
-          results_container = containers.OsqueryResult(
-              name=name,
-              description=description,
-              query=query,
-              hostname=hostname,
-              data_frame=pd.DataFrame(),
-              flow_identifier=flow_identifier,
-              client_identifier=client_identifier)
-          self.state.StoreContainer(results_container)
-          continue
-
-        for data_frame in results:
-          self.logger.info(
-              f'{str(flow_id)} ({container.hostname}): {len(data_frame)} rows '
-              'collected')
-
-          dataframe_container = containers.OsqueryResult(
-              name=name,
-              description=description,
-              query=query,
-              hostname=hostname,
-              data_frame=data_frame,
-              flow_identifier=flow_identifier,
-              client_identifier=client_identifier)
-
-          self.state.StoreContainer(dataframe_container)
+        host_osquery_future = executor.submit(
+          self._ProcessQuery, container.hostname, osquery_container)
+        host_osquery_futures.append(host_osquery_future)
+
+    for host_osquery_future in host_osquery_futures:
+      if host_osquery_future.exception():
+        raise DFTimewolfError('Error with osquery future') \
+            from host_osquery_future.exception()
```

Comment:
Please use `self.ModuleError` or `self.logger.error/warning` depending on how you want the execution flow of your recipe to go.  Also, let's try to keep code-related jargon (e.g. `future`) out of the error messages which are for the user.

Reply:
Done